hxprune.1
.TH "HXPRUNE" "1" "10 Jul 2011" "7.x" "HTML-XML-utils"
.SH NAME
hxprune \- remove marked elements from an HTML file
.SH SYNOPSIS
.B hxprune
.RB "[\| " \-c
.IR class " \|]"
.RB "[\| " \-x " \|]"
.RI "[\| " file " \|]"
.SH DESCRIPTION
.B hxprune
removes specially marked elements (with their content and children)
from an HTML file. By default it removes elements with a class
attribute of "exclude". The class attribute is parsed as a
whitespace-separated list of words and if one of those words
is "exclude", the element is not copied to the output.
.SH OPTIONS
The following options are supported:
.TP
.BI \-c " class"
Specify a different
.I class
than "exclude".
.TP
.B \-x
Force output to be in XML-style, i.e., without short attributes and
with a slash (/) to mark empty elements.
.\" TODO


ChangeLog
2019-10-05  Bert Bos  <bert@w3.org>

	* Published version 7.8.

	* textwrap.c, langinfo.c, hxnormalize.c: Added knowledge about
	languages that do not use spaces between words. In such languages,
	a newline should not be converted to a space in outc() in
	textwrap.c, but only to a break opportunity.

2019-08-28  Bert Bos  <bert@w3.org>

	* hxtoc.c: The element to group headings in HTML5 is called
	HGROUP, not HEADER. The heading of a section (SECTION, ARTICLE,
	etc.) need not be the first element, there may be non-header
	elements before it.

	* hxwls.c: Print "longdesc", "classid" or "codebase" in the second
	column for the corresponding attribute. Also recognize srcset
	(somewhat).

2018-08-03  Bert Bos  <bert@w3.org>

	* hxnormalize.c: Added option -X to indicate the input is XML
	instead of HTML. Handle conversion of CDATA elements to XML by
	escaping < and & instead of adding <![CDATA[. Added corresponding
	test normalize13.sh.

2018-04-29  Bert Bos  <bert@w3.org>

	* Released version 7.7.

	* dtd.hash: Don't include the arguments in the forward declaration
	of lookup_element(), because those arguments differ slightly
	depending on which version of gperf is used to generate dtd.c:
	"unsigned int" in gperf 3.0 vs "size_t" in gperf 3.1.

	* conficgure.ac: Check for libiconv with the AM_ICONV macro (from
	gettext) instead of AC_SEARCH_LIBS. (Advice from Christian
	Weisgerber, received via Anthony Bentley). That in turn requires
	adding config.rpath to the EXTRA_DIST in Makefile.am.

2018-02-23  Bert Bos  <bert@w3.org>

	* tests/index10.sh: Added a test for a term database.

	* hxindex.c: Replaced fgets() by getline(), to allow lines of
	arbitrary length in the term database.

2018-02-15  Bert Bos  <bert@w3.org>

	* tests/index9.sh: Added a test for hxindex's option -N.

	* tests/index5.sh: Updated for hxindex's option -N.

	* tests/cite5.sh: Added the example from hxmkbib.1 as a test.

	* tests/normalize12.sh: Added a test for the NOSCRIPT element.

	* hxindex.1: Added option -N.

	* hxmkbib.1: The example was missing some "%" signs.

	* hxmkdir.c: Read the template before reading the auxfile, so that
	a pipe like "hxcite -a auxfile bib template | hxmkbib -a auxfile
	bib" becomes possible. hxcite will have the time to finishe
	writing the auxfile before hxmkbib reads it.

2018-02-12  Bert Bos  <bert@w3.org>

	* hxindex.c: Added option -N.

	* dtd.hash: Since HTML5, the NOSCRIPT element is phrase-level
	instead of block-level.

2018-02-09  Bert Bos  <bert@w3.org>

	* Version 7.6 released.

	* selmatch.c: matches_sel() could read uninitialized memory
	(h->parent where h is the root of a tree) potentially leading to a
	SIGBUS. Thanks to Anthony J. Bentley for finding this bug.

2018-01-20  Bert Bos  <bert@w3.org>

	* Version 7.5 released.

	* hxcite.1: Spelling error.

	* hxincl.1: Spelling error.

	* url.c: Fixed unsafe way to call warnx(). Thanks to Robin
	Naundorf for these three corrections.

2017-12-08  Bert Bos  <bert@w3.org>

	* tests/normalize11.sh: Added to test for "main".

	* dtd.hash: Added the HTML5 element "main".

2017-12-07  Bert Bos  <bert@w3.org>

	* hxwls.c: Added support for HTML5 elements (source, audio, video)
	and for the classid and codebase attribute. Fixed an error: input
	has a src attribute, not href.

2017-11-24  Bert Bos  <bert@w3.org>

	* Version 7.4 released.

	* hxremove.1, hxselect.1: Updated.

	* tests/remove4.sh, tests/select18.sh, tests/select19.sh,
	tests/select20.sh, tests/select21.sh, tests/select22.sh,
	tests/remove5.sh tests/remove6.sh tests/remove7.sh,
	tests/select23.sh tests/select24.sh: Added.

	* selmatch.c: Now handles the ':not()' pseudo-class and
	comma-separated selectors. Fixed a bug in matching the ':root'
	pseudo-class.

	* hxremove.c, hxselect.c: parse_selector() has different
	arguments.

	* selector.c: Now parses the ':not()' pseudo-class and
	comma-separated selectors. Errors are now printed with a newline
	at the end. Added new functions to dump (serialize) a selector to
	a file, useful for debugging.

	* openurl.c: Functions now either return CURLcode or CURLMcode,
	but not some mixture.

2017-11-11  Bert Bos  <bert@w3.org>

	* dtd.hash: Added some HTML5 elements: source, track, wbr (because
	they are empty) and audio and video (as parents for source and
	track).

	* toc4.sh, normalize10.sh: Added tests for the <source> element.

	* Released version 7.3.

2017-10-16  Bert Bos  <bert@w3.org>

	* normalize9.sh: Added test for A inside FIGCAPTION.

	* dtd.hash: An A can now be inside a FIGCAPTION. Made source
	shorter and more readable with some macros.

2017-10-06  Bert Bos  <bert@w3.org>

	* Released version 7.2.

	* tests/extract1.sh: Added option -N to nc, to close the
	connection after EOF on stdin, because hxextract (with libcurl)
	seems to wait for the closing of the connection.

	* tests/unent1.ssh, tests/unent2.ssh, tests/unent3.ssh,
	tests/unent4.ssh, tests/unent5.ssh: Added tests for hxunent.

	* hxunentmain.c: Fixed bug reported by Laurent Martelli: When a
	line was longer than the 4096-byte buffer, an entity could be on
	the edge between two buffers and not be recognized.

2017-08-13  Bert Bos  <bert@w3.org>

	* extract1.sh: return code 77 (skip) rather than 0 when netcat is
	unavailable.

	* tests/pipe4.sh, tests/select01.sh, ..., tests/select17.sh,
	tests/unpipe6.sh: New tests.

	* scan.l: Accept non-ASCII characters in names (element names,
	attribute names, etc.)

	* hxselect.c, hxremove.c: Put their common code in
	selmatch.c. hxselect now uses the data structures from tree.c,
	just like hxremove.c. The non-element selector '::attr(...)' is
	now also supported.

	* selector.c, selmatch.c: Added support for more structural
	pseudo-classes (last-child, nth-last-of-type, etc.). Generalized
	support for backslash escapes. Support hexadecimal backslash
	escapes also outside ASCII range (such characters are converted to
	UTF-8). Bug fixes.

2016-10-17  Bert Bos  <bert@w3.org>

	* Released version 7.1.


hxprune.c
/*
 * Remove subtrees which have a certain class attribute.
 *
 * Copyright © 1994-2000 World Wide Web Consortium
 * See http://www.w3.org/Consortium/Legal/copyright-software
 *
 * Bert Bos <bert@w3.org>
 * Created Feb 2000
 * $Id: hxprune.c,v 1.8 2017/11/24 09:50:25 bbos Exp $
 *
 **/
/* #include <mcheck.h> */
#include "config.h"
#include <assert.h>
#include <ctype.h>
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <stdbool.h>
#if STDC_HEADERS
# include <string.h>
#else
# ifndef HAVE_STRCHR
#  define strchr index
#  define strrchr rindex
# endif
# ifndef HAVE_STRSTR
#  include "strstr.e"
# endif
#endif
#include "export.h"
#include "types.e"
#include "tree.e"
#include "html.e"
#include "scan.e"
#include "dict.e"
#include "openurl.e"
#include "class.e"
#include "errexit.e"

#define EXCLUDE_CLASS "exclude"			/* Default value for class */

static Tree tree;
static bool xml = false;			/* Use <empty /> convention */


/* handle_error -- called when a parse error occurred */
static void handle_error(void *clientdata, const string s, int lineno)
{
  fprintf(stderr, "%d: %s\n", lineno, s);
}

/* start -- called before the first event is reported */
static void* start(void)
{
  tree = create();
  return NULL;
}
  
/* end -- called after the last event is reported */
static void end(void *clientdata)
{
  /* skip */
}

/* handle_comment -- called after a comment is parsed */
static void handle_comment(void *clientdata, string commenttext)
{
  tree = append_comment(tree, commenttext);
}

/* handle_text -- called after a tex chunk is parsed */
static void handle_text(void *clientdata, string text)
{
  tree = append_text(tree, text);
}

/* handle_declaration -- called after a declaration is parsed */
static void handle_decl(void *clientdata, string gi,
			string fpi, string url)
{
  tree = append_declaration(tree, gi, fpi, url);
}

/* handle_proc_instr -- called after a PI is parsed */
static void handle_pi(void *clientdata, string pi_text)
{
  tree = append_procins(tree, pi_text);
}

/* handle_starttag -- called after a start tag is parsed */
static void handle_starttag(void *clientdata, string name, pairlist attribs)
{
  tree = html_push(tree, name, attribs);
}

/* handle_emptytag -- called after an empty tag is parsed */
static void handle_emptytag(void *clientdata, string name, pairlist attribs)
{
  tree = html_push(tree, name, attribs);
}

/* handle_pop -- called after an endtag is parsed (name may be "") */
static void handle_endtag(void *clientdata, string name)
{
  tree = html_pop(tree, name);
}

/* prune -- write the tree, suppressing elements with a certain class */
static void prune(Tree t, const string class)
{
  Tree h;
  pairlist a;

  for (h = t->children; h != NULL; h = h->sister) {
    switch (h->tp) {
      case Text:
	printf("%s", h->text);
	break;
      case Comment:
	printf("<!--%s-->", h->text);
	break;
      case Declaration:
	printf("<!DOCTYPE %s", h->name);
	if (h->text) printf(" PUBLIC \"%s\"", h->text);
	if (h->url) printf(" %s\"%s\"", h->text ? "" : "SYSTEM ", h->url);
	printf(">");
	break;
      case Procins:
	printf("<?%s>", h->text);
	break;
      case Element:
	if (! has_class(h->attribs, class)) {
	  printf("<%s", h->name);
	  for (a = h->attribs; a != NULL; a = a->next) {
	    printf(" %s", a->name);
	    if (a->value != NULL) printf("=\"%s\"", a->value);
	    else if (xml) printf("=\"%s\"", a->name);
	  }
	  if (is_empty(h->name)) {
	    assert(h->children == NULL);
	    printf(xml ? " />" : ">");
	  } else {
	    printf(">");
	    prune(h, class);
	    printf("</%s>", h->name);
	  }
	}
	break;
      case Root:
	assert(! "Cannot happen");
	break;
      default:
	assert(! "Cannot happen");
    }
  }
}

/* usage -- print usage message and exit */
static void usage(string name)
{
  fprintf(stderr, "Usage: %s [-c class] [-x] [html-file]\n", name);
  exit(1);
}


int main(int argc, char *argv[])
{
  int i, status;
  string class = EXCLUDE_CLASS;

  /*  mtrace(); */

  /* Bind the parser callback routines to our handlers */
  set_error_handler(handle_error);
  set_start_handler(start);
  set_end_handler(end);
  set_comment_handler(handle_comment);
  set_text_handler(handle_text);
  set_decl_handler(handle_decl);
  set_pi_handler(handle_pi);
  set_starttag_handler(handle_starttag);
  set_emptytag_handler(handle_emptytag);
  set_endtag_handler(handle_endtag);

  yyin = stdin;
  for (i = 1; i < argc; i++) {
    if (eq(argv[i], "-c")) {
      if (i >= argc - 1) usage(argv[0]);
      class = argv[++i];
    } else if (eq(argv[i], "-x")) {
      xml = true;
    } else {
      yyin = fopenurl(argv[i], "r", &status);
      if (yyin == NULL) {perror(argv[1]); exit(2);}
      if (status != 200) errexit("%s : %s\n", argv[i], http_strerror(status));
    }
  }

  if (yyparse() != 0) {
    exit(3);
  }
  tree = get_root(tree);
  prune(tree, class);
  tree_delete(tree);				/* Just to test memory mgmt */
  return 0;
}
